### DISCUSSION TITLE
Empty GITHUB_ENV variables on PowerShell

### DISCUSSION THREAD

**Discussion by luigiberrettini** [community/community#26641] [Actions]
<p>I am running a job on <code>windows-latest</code> and after setting a variable in a step it is empty in the following ones:</p>
<pre><code class="lang-auto"># STEP 1
Write-Host "workingDir=$workingDir" |
    Out-File -FilePath $env:GITHUB_ENV -Encoding utf8 -Append

# STEP 2
Write-Host "env:workingDir '$env:workingDir'"
</code></pre>
<p>I tried without specifying the shell and the step uses PowerShell 7.0:<br>
<code>C:\Program Files\PowerShell\7\pwsh.EXE -command ". '{0}'"</code></p>
<p>Then I tried specifying <code>shell: powershell</code> and the step uses Windows PowerShell 5.1:<br>
<code>C:\windows\System32\WindowsPowerShell\v1.0\powershell.EXE -Command "&amp; '{0}'"</code></p>
<p>Finally I tried specifying <code>shell: pwsh -Command "&amp; '{0}'"</code> and the step uses PowerShell 7.0:<br>
<code>C:\Program Files\PowerShell\7\pwsh.EXE -Command "&amp; '{0}'"</code></p>
<p>It would be great if someone could help me find out what I am doing wrong.</p>
<p>Thanks!</p>

**Comment by luigiberrettini** [comment, marked as answer]
<p>The Environment files <a href="https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#environment-files" rel="noopener nofollow ugc">docs</a> use <code>echo</code> in PowerShell.</p>
<p>After some investigation I found out that <code>echo</code> is an alias for <code>Write-Output</code> not <code>Write-Host</code> which only outputs to the console, not to stdout.</p>
<p>I ended up using <code>Add-Content</code> which uses UTF8 (without BOM) by default:</p>
<pre><code class="lang-auto">Add-Content -Path $env:GITHUB_ENV -Value "name=$value"
</code></pre>

**Comment by Simran-B** [comment]
<aside class="quote no-group" data-username="luigiberrettini" data-post="2" data-topic="147626">
<div class="title">
<div class="quote-controls"></div>
 luigiberrettini:</div>
<blockquote>
<p>I ended up using <code>Add-Content</code> which uses UTF8 (without BOM) by default</p>
</blockquote>
</aside>
<p>This is not the case in PowerShell, only in PowerShell Core (pwsh)! If we assume pwsh, then you can simplify it to this:</p>
<p><code>echo "{name}={value}" &gt;&gt; $Env:GITHUB_ENV</code></p>
<p>For maximum compatibility, you may go with the docs recommendation:</p>
<p><code>echo "{name}={value}" | Out-File -FilePath $Env:GITHUB_PATH -Encoding utf8 -Append</code></p>

**Comment by luigiberrettini** [comment]
<p>For maximum compatibility it is enough to add <code>-Encoding UTF8</code> to <code>Add-Content</code> which avoids piping and using the <code>echo</code> alias (non idiomatic).</p>

**Comment by Simran-B** [comment]
<p>Shouldn’t it be <code>-Encoding utf8NoBOM</code>?</p>

**Comment by luigiberrettini** [comment]
<p>Not if you take into consideration the docs recommendation and the absence of <code>utf8NoBom</code> from Windows PowerShell 5.1 (<code>powershell</code>)</p>

**Comment by Simran-B** [comment]
<p>I see. Just checked locally that <code>Add-Content</code> doesn’t write a BOM if the target file does not exist yet. It does add a line break (<code>\r\n</code>) automatically, but so does <code>Out-File</code>. What’s problematic is that <code>Out-File</code> writes a BOM if it’s a new file and that there is no <code>utf8NoBOM</code> encoding available.</p>
<p>FWIW, GitHub-hosted Windows runners provide the following PS versions:</p>
<div class="md-table">
<table>
<thead>
<tr>
<th></th>
<th><code>powershell</code></th>
<th><code>pwsh</code></th>
</tr>
</thead>
<tbody>
<tr>
<td><strong><code>windows-2016</code></strong></td>
<td>5.1.14393.3866</td>
<td>7.1.0</td>
</tr>
<tr>
<td><strong><code>windows-2019</code></strong></td>
<td>5.1.17763.1490</td>
<td>7.1.0</td>
</tr>
</tbody>
</table>
</div>

**Comment by davesbrown** [comment]
<p>I’ve tried all of the above, but I’m still getting empty variable for $BUILD_VERSION</p>
<pre><code>runs-on: windows-2019
env:
  BRANCH_NAME: ${{ github.ref_name }}
  
steps:

- uses: actions/checkout@v3

- name: Create Variables
  run: |
    $sha_short = (git rev-parse --short HEAD)
    $BUILD_VERSION = (get-date -format "yyyy.MM.dd") + '-' + $sha_short
    Write-Output "Setting BuildNumber = $BUILD_VERSION"
    Write-Output "Branch = $env:BRANCH_NAME"
    #echo "BUILD_VERSION=$BuildNumber" | Out-File -FilePath $Env:GITHUB_ENV -Encoding utf-8 -Append
    Add-Content -Path $env:GITHUB_ENV -Value "name=$BUILD_VERSION"
    #echo BUILD_VERSION=$BuildNumber &gt;&gt; $GITHUB_ENV
  shell: pwsh
    
- name: Echo Variables
  run: |
    echo "${{ env.BUILD_VERSION }}"
    echo "${{ env.BRANCH_NAME }}"</code></pre>

**Comment by Simran-B** [comment]
<p>None of the three variants in your example looks right:</p>
<ul>
<li>
<p><code>echo "BUILD_VERSION=$BuildNumber" | Out-File -FilePath $Env:GITHUB_ENV -Encoding utf-8 -Append</code><br>
The name <code>BUILD_VERSION</code> is correct, but you never defined a variable <code>$BuildNumber</code></p>
</li>
<li>
<p><code>Add-Content -Path $env:GITHUB_ENV -Value "name=$BUILD_VERSION"</code><br>
You correctly reference the <code>$BUILD_VERSION</code> variable, but the name you give the environment variable is <code>name</code></p>
</li>
<li>
<p><code>echo BUILD_VERSION=$BuildNumber &gt;&gt; $GITHUB_ENV</code><br>
Again, you reference <code>$BuildNumber</code> which was never defined, and also <code>$GITHUB_ENV</code> instead of <code>$Env:GITHUB_ENV</code>.</p>
</li>
</ul>
<p>Try the following:</p>
<pre><code class="lang-auto">echo "BUILD_VERSION=$BUILD_VERSION" &gt;&gt; $Env:GITHUB_ENV
</code></pre>

**Comment by davesbrown** [comment]
<p>Hi Simran-B,</p>
<p>Thanks, after too many hours looking over it and trying different things, your fresh eyes helped solve.</p>
